### PR TITLE
[DPE-6371] Add TLS support

### PR DESCRIPTION
**Pull request by imanenami**
# Changes:
- Adds TLS support on the REST interface.
- Adds TLS support on the `kafka-client` interface

**Note**: There is an issue with `kafka-operator` when TLS is enabled on `kafka-client` interface, reported [here](https://github.com/canonical/kafka-operator/issues/300), which affects this PR. once that issue is fixed, the `kafka-client + TLS` should work properly on this charm.

### CODE REVIEW

File: src/core/models.py [hunk 1]
```diff
@@ -163,12 +224,18 @@ def __init__(
...
 
     @property
     def unit_id(self) -> int:
         """The id of the unit from the unit name."""
         return int(self.unit.name.split("/")[1])
 
+    @property
+    def tls(self) -> TLSContext:
+        """TLS Context of the worker unit."""
+        return self._tls
```
Comment (deusebio) [inline src/core/models.py:237]: **question, suggesiton** why not just

```
@property
    def tls(self) -> TLSContext:
        """TLS Context of the worker unit."""
        return TLSContext(relation, data_interface, component)
```

if you are afraid of re-instantiation that that could be an issue, you could also use the cached_property decorator. 

I don't know defining a kinda hidden attribute to be referenced in a property looks kinda weird to me, but it is not a strong point

File: src/literals.py [hunk 2]
```diff
@@ -53,6 +57,19 @@
...
+    CA = "ca"
+    CHAIN = "chain"
+    CERT = "certificate"
+    CSR = "csr"
+    BROKER_CA = "broker"
+    PRIVATE_KEY = "private-key"
+    KEYSTORE_PASSWORD = "keystore-password"
+    TRUSTSTORE_PASSWORD = "truststore-password"
+    KEYS = {CA, CERT, CSR, CHAIN}
+    SECRETS = [CERT, CSR, PRIVATE_KEY, KEYSTORE_PASSWORD, TRUSTSTORE_PASSWORD]
```
Comment (deusebio) [inline src/literals.py:70]: **praise, thought** I really like that we define the string constants in a class, to keep them all together. I'm just wondering whether the `literals` module is the right place. This would pertain to the logic to parse databags into models, if we had a parsers module, this would seem appropriate there, but until then, to me, `core/models` would seem a better place (since that is already dealing with serialization/deserialization of databags)

File: tests/integration/test_tls.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        )
+
+    with pytest.raises(ConnectionError):
+        _ = await make_connect_api_request(ops_test, proto="http")
+
+    response = await make_connect_api_request(ops_test, proto="https", verify=False)
+
+    assert response.status_code == 200
+
+    cert = await get_certificate(ops_test)
```
Comment (deusebio) [inline tests/integration/test_tls.py:71]: **todo** beside doing the indirect check by downloading the cert, we could actually also do the direct check by connecting to the endpoint. @welpaolo was doing this in the integration tests for kyuubi, see [here](https://github.com/canonical/kyuubi-k8s-operator/pull/47/files#diff-d0fbf30f9fc8d4e2b961cdc1adf45385c53e4e091c3de0cdd2a1f882366ac444R140). Maybe we could do something similar here too. That would be the best check to make sure that the correct certificate is presented to clients.

**Review by deusebio — COMMENTED** [review]
The code looks very very good! I feel that we really do have a lot of re-usable code and patterns for TLS that we can use across the board. Reviewing these PRs is becoming almost boring :P I believe the next step here would really be to decouple the classes into a common class, and then implement the specifics bit in a subclass. It is similar to a comment I had to a PR in Spark, see [here](https://github.com/canonical/kyuubi-k8s-operator/pull/47#discussion_r1889839251). Maybe you can pair between teams to discuss this

Also by taking inspiration from the PR in Spark, I only have a small point to add a "direct" check for the certificate in the integration tests.

File: tests/integration/test_tls.py (hunk 3 above)
Comment (imanenami) [reply tests/integration/test_tls.py:72]: Great point, made the changes here: https://github.com/canonical/kafka-connect-operator/pull/4/commits/55a0e902f15827769d155c18818bea1f45ebb4e1

File: src/core/models.py (hunk 1 above)
Comment (imanenami) [reply src/core/models.py]: Makes sense, fixed: https://github.com/canonical/kafka-connect-operator/pull/4/commits/bc09ff5f733dd3d1638de6de228f64a9c89ed729

**Review by deusebio — APPROVED** [review]
Looks good! thank you for addressing the comments

File: src/core/models.py [hunk 4]
```diff
@@ -169,6 +240,11 @@ def unit_id(self) -> int:
         """The id of the unit from the unit name."""
         return int(self.unit.name.split("/")[1])
 
+    @property
+    def tls(self) -> TLSContext:
+        """TLS Context of the worker unit."""
+        return TLSContext(self.relation, self.data_interface, self.component)
```
Comment (marcoppenheimer) [inline src/core/models.py:246]: **praise:** This is such an improvement in design. If you ever get REALLY bored, I would love to see you go through Kafka-kafka with a fine-toothed comb and follow some of the learnings here.

File: src/events/kafka.py [hunk 5]
```diff
@@ -55,6 +55,20 @@ def _on_relation_created(self, event: RelationCreatedEvent) -> None:
...
+                self.charm.context.worker_unit.update(
+                    {tls_context.TRUSTSTORE_PASSWORD: self.charm.workload.generate_password()}
+                )
+
+            self.charm.tls_manager.import_cert(
+                tls_context.BROKER_CA,
+                f"{tls_context.BROKER_CA}.pem",
+                cert_content=self.context.kafka_client.broker_ca,
+            )
+
```
Comment (marcoppenheimer) [inline src/events/kafka.py:71]: **question:** Am I correct in thinking that this currently does nothing, but in the future will be used?

File: src/events/tls.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        if not self.unit_tls_context.keystore_password:
+            self.charm.context.worker_unit.update(
+                {self.unit_tls_context.KEYSTORE_PASSWORD: self.charm.workload.generate_password()}
+            )
+        if not self.unit_tls_context.truststore_password:
+            self.charm.context.worker_unit.update(
+                {
+                    self.unit_tls_context.TRUSTSTORE_PASSWORD: self.charm.workload.generate_password()
+                }
+            )
```
Comment (marcoppenheimer) [inline src/events/tls.py:82]: **suggestion:** Consider something similar to:
```python
self.charm.context.worker_unit.update(
    {
        self.unit_tls_context.PRIVATE_KEY: self.unit_tls_context.private_key
        or generate_private_key().decode("utf-8"),
        self.unit_tls_context.KEYSTORE_PASSWORD: self.unit_tls_context.keystore_password
        or self.charm.workload.generate_password(),
        self.unit_tls_context.TRUSTSTORE_PASSWORD: self.unit_tls_context.truststore_password
        or self.charm.workload.generate_password(),
    }
)
```

Formats like crap because it's a bit verbose, so it's up to you.

File: src/events/connect.py [hunk 7]
```diff
@@ -45,14 +46,12 @@ def _on_install(self, event: InstallEvent) -> None:
...
-            if self.context.ready:
-                self.charm._set_status(Status.SERVICE_NOT_RUNNING)
-            else:
-                self.charm._set_status(self.context.status)
+        elif self.context.ready:
+            self.charm._set_status(Status.SERVICE_NOT_RUNNING)
+        else:
+            self.charm._set_status(self.context.status)
 
     def _on_config_changed(self, event: ConfigChangedEvent):
```
Comment (marcoppenheimer) [inline src/events/connect.py:56]: **todo:** Here you want to add something similar to what we have in Kafka, to ensure that we request new certificates when our SANs change (if IP changed during a pod reschedule, for example). Example from Kafka:
```python
        current_sans = self.tls_manager.get_current_sans()
        if not (properties and zk_jaas):
            # Event fired before charm has properly started
            event.defer()
            return
        current_sans_ip = set(current_sans["sans_ip"]) if current_sans else set()
        expected_sans_ip = set(self.tls_manager.build_sans()["sans_ip"]) if current_sans else set()
        sans_ip_changed = current_sans_ip ^ expected_sans_ip
        current_sans_dns = set(current_sans["sans_dns"]) if current_sans else set()
        expected_sans_dns = (
            set(self.tls_manager.build_sans()["sans_dns"]) if current_sans else set()
        )

        sans_dns_changed = (current_sans_dns ^ expected_sans_dns) - {
            # we omit 'kafka/{unit_id}' and 'kafka' here to avoid a bug with Digicert not supporting '/' characters in SANs
            # Digicert truncates the 'kafka/{unit_id}' to just 'kafka'
            # i.e don't assume we need new certs if 'diff' includes those value, as these SANs aren't typically used anyway
            self.charm.state.unit_broker.unit.name,
            self.charm.state.cluster.app.name,
        }

        # update environment
        self.config_manager.set_environment()

    
          
            
    

          
          Expand Down
    
    
  
        self.charm.unit.set_workload_version(self.workload.get_version())
        if sans_ip_changed or sans_dns_changed:
            logger.info(
                (
                    f'Broker {self.charm.unit.name.split("/")[1]} updating certificate SANs - '
                    f"OLD SANs IP = {current_sans_ip - expected_sans_ip}, "
                    f"NEW SANs IP = {expected_sans_ip - current_sans_ip}, "
                    f"OLD SANs DNS = {current_sans_dns - expected_sans_dns}, "
                    f"NEW SANs DNS = {expected_sans_dns - current_sans_dns}"
                )
            )
            self.charm.tls.certificates.on.certificate_expiring.emit(
                certificate=self.charm.state.unit_broker.certificate,
                expiry=datetime.now().isoformat(),
            )  # new cert will eventually be dynamically loaded by the broker
            self.charm.state.unit_broker.update(
                {"certificate": ""}
            )  # ensures only single requested new certs, will be replaced on new certificate-available event
return # early return here to ensure new node cert arrives before updating advertised.listeners
```

Comment (marcoppenheimer) [reply src/events/connect.py:56]: That paste was awful. It's in https://github.com/canonical/kafka-operator/pull/297

File: src/managers/tls.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+        context: Context,
+        workload: WorkloadBase,
+        substrate: Substrates,
+    ):
+        self.unit_context: WorkerUnitContext = context.worker_unit
+        self.tls_context: TLSContext = self.unit_context.tls
+        self.workload = workload
+        self.substrate = substrate
+
+    @cached_property
```
Comment (marcoppenheimer) [inline src/managers/tls.py:39]: **question:** Any reason this is cached? It seems to just be a string. Unless we're going for uber performance 😄

File: src/managers/tls.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        )
+
+    def set_chain(self) -> None:
+        """Sets the unit chain."""
+        if not self.tls_context.chain:
+            logger.error("Can't set chain to unit, missing chain in relation data")
+            return
+
+        for i, chain_cert in enumerate(self.tls_context.chain):
+            self.workload.write(content=chain_cert, path=f"{CONFIG_DIR}/chain{i}.pem")
```
Comment (marcoppenheimer) [inline src/managers/tls.py:85]: **praise:** Thanks for porting this.

**Review by marcoppenheimer — APPROVED** [review]
Outstanding work - though I think you're a victim of success here. I mentioned in a comment already, but I think that a lot of the small tasteful changes you've made here, really should be ported back to Kafka. I think this repo is shaping up to be the example that we should follow in Kafka.

File: src/events/kafka.py (hunk 5 above)
Comment (imanenami) [reply src/events/kafka.py:71]: Exactly!

File: src/managers/tls.py (hunk 8 above)
Comment (imanenami) [reply src/managers/tls.py:39]: Skips a bunch of conditional checks on hooks like `certificate-available`, nothing important :)

File: src/events/tls.py (hunk 6 above)
Comment (imanenami) [reply src/events/tls.py:82]: Loved it, done :+1:

**Comment by imanenami** [comment]
@deusebio @marcoppenheimer 
I had to rebase on `main` and sign the commits, and I have made the necessary changes for supporting SANs change detection. I'll really appreciate a re-review. The changes after your first approval is on this commit: https://github.com/canonical/kafka-connect-operator/pull/4/commits/a0de51c5359beca42d55a062a74805d60da1cced

File: src/managers/tls.py [hunk 10]
```diff
@@ -175,7 +179,35 @@ def get_current_sans(self) -> Sans | None:
...
+        logger.info(
+            (
+                f"SANs change detected - "
+                f"OLD SANs IP = {current_sans_ip - expected_sans_ip}, "
+                f"NEW SANs IP = {expected_sans_ip - current_sans_ip}, "
+                f"OLD SANs DNS = {current_sans_dns - expected_sans_dns}, "
+                f"NEW SANs DNS = {expected_sans_dns - current_sans_dns}"
+            )
+        )
+        return True
```
Comment (marcoppenheimer) [inline src/managers/tls.py:210]: **praise:** As always, better than before, moving this to a manager.

**Review by marcoppenheimer — APPROVED** [review]

File: src/events/connect.py (hunk 7 above)
Comment (deusebio) [reply src/events/connect.py]: **question** how would the SANs change here? I mean, pod rescheduling would always re-request certificates I would think, no? it is like a new unit coming up live... We don't have extra listeners / sans or load-balancing yet, right?

Having said this, probably have the code to be already structured when the IP may change is not bad, I'm just wondering whether we would already have a use-case (and therefore a test for this)

Comment (imanenami) [reply src/events/connect.py]: To test the logic, I've already added a unit test. We could add an integration test by simulating an IP change with `lxc config device add {machine_name} eth0 none` and then `lxc config device remove {machine_name} eth0` but I think it's more applicable to k8s ...

Comment (marcoppenheimer) [reply src/events/connect.py]: If the pod reschedules, it's still the same 'unit', so it would grab the existing certs from relation data. But the IP would be different now.

Comment (marcoppenheimer) [reply src/events/connect.py]: Though come to think of it, we should probably add the `expose-external` logic here. I'll put it in the next sprint.

**Review by deusebio — APPROVED** [review]
Looks good! Great job!

**Comment by imanenami** [comment]
@deusebio @marcoppenheimer 

I had to rebase with main and apply the fix described in [DA-152](https://docs.google.com/document/d/1umFlaBvYAfCkUf5JphuveIFwuqEJzwKEmWQR7vqHVH0) for cert chains. Also made two small fixes regarding installing status & redundant wokrer restarts when attaching plugins manually

**Review by marcoppenheimer — APPROVED** [review]

**Review by deusebio — APPROVED** [review]
Looks good!

**Comment by imanenami** [comment]
Thanks! signing commits, no change: https://github.com/canonical/kafka-connect-operator/compare/1fb507b8a67122b65e26bd83f33a6a3a27b15a15..aaf2f7275ce1e8256295f4543b682768a02dbdf7
